### ISSUE TITLE
'Generate GetHashCode' generates a call to ValueType.GetHashCode

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#26141]
VS 15.6.6.

If you use Roslyn's 'Generate GetHashCode...' suggestion on a struct, it asks you which properties to include, and then it generates code combining those with an additional hash, `base.GetHashCode()`:

```cs
struct Foo
{
    public Type FixtureType { get; }
    public MethodInfo Method { get; }

    public override int GetHashCode()
    {
        var hashCode = 662238274;
        hashCode = hashCode * -1521134295 + base.GetHashCode();
        hashCode = hashCode * -1521134295 + EqualityComparer<Type>.Default.GetHashCode(FixtureType);
        hashCode = hashCode * -1521134295 + EqualityComparer<MethodInfo>.Default.GetHashCode(Method);
        return hashCode;
    }
}
```

Just like when inheriting from `System.Object`, inheritors from `System.ValueType` are always going to want to replace the base has calculation not combine it. Not only can this easily result in incorrect hashes which contradict the `Equals` implementation, many times the entire purpose of overriding `ValueType.GetHashCode` is to provide a better-performing implementation.

**Comment by jnm2** [comment]
So, stepping back, if you've ever used (a recent?) Roslyn to generate GetHashCode for a struct which uses non-default equality, you have a bug to go fix. The generated implementation breaks the invariant that if Equals returns true for two objects, their hash codes should be equal.

**Comment by jnm2** [comment]
It looks like @CyrusNajmabadi has done something here which made it into master two months ago. Think the problem is already fixed in master?

https://github.com/dotnet/roslyn/pull/24161/files#diff-a0d90f84d4625c53558ee000fb8ed2bbR147

**Comment by jmarolf** [comment]
@jnm2 this problem should already be fixed in the latest Visual Studio 15.7 Preview

If you generate hashcode + equals on the following  struct
```C#
using System;
using System.Collections.Generic;
using System.Reflection;

struct Foo
{
    public Type FixtureType { get; }
    public MethodInfo Method { get; }
}
```

you get this:

```C#
using System;
using System.Collections.Generic;
using System.Reflection;

struct Foo
{
    public Type FixtureType { get; }
    public MethodInfo Method { get; }

    public override bool Equals(object obj)
    {
        if (!(obj is Foo))
        {
            return false;
        }

        var foo = (Foo)obj;
        return EqualityComparer<Type>.Default.Equals(FixtureType, foo.FixtureType) &&
               EqualityComparer<MethodInfo>.Default.Equals(Method, foo.Method);
    }

    public override int GetHashCode()
    {
        var hashCode = 662238274;
        hashCode = hashCode * -1521134295 + EqualityComparer<Type>.Default.GetHashCode(FixtureType);
        hashCode = hashCode * -1521134295 + EqualityComparer<MethodInfo>.Default.GetHashCode(Method);
        return hashCode;
    }
}
```

**Comment by jnm2** [comment]
Closing as fixed in 15.7.

**Comment by CyrusNajmabadi** [comment]
This was fixed with https://github.com/dotnet/roslyn/pull/24161
